From: Simon McVittie <smcv@debian.org>
Date: Wed, 13 Dec 2023 10:34:02 +0000
Subject: Mark several additional tests as flaky

Most of these have never been reliable on Debian buildds, especially
on ARM, so having them fail is not a regression. They usually succeed
after a few retries, but that's time-consuming and unsuitable to be a
QA gate.

Signed-off-by: Simon McVittie <smcv@debian.org>
Forwarded: no
---
 src/tests/clutter/conform/gesture.c | 24 ++++++++++++++++++++++++
 src/tests/clutter/conform/grab.c    |  6 ++++++
 src/tests/meson.build               | 11 +++++++++--
 3 files changed, 39 insertions(+), 2 deletions(-)

diff --git a/src/tests/clutter/conform/gesture.c b/src/tests/clutter/conform/gesture.c
index f3901a2..c0affef 100644
--- a/src/tests/clutter/conform/gesture.c
+++ b/src/tests/clutter/conform/gesture.c
@@ -133,6 +133,12 @@ gesture_state_machine_move_to_waiting (void)
   ClutterGesture *gesture = CLUTTER_GESTURE (g_object_new (TEST_TYPE_GESTURE, NULL));
   gboolean was_updated;
 
+  if (g_getenv ("DEB_ALLOW_FLAKY_TESTS") == NULL)
+    {
+      g_test_skip ("https://gitlab.gnome.org/GNOME/mutter/-/issues/3521");
+      return;
+    }
+
   virtual_pointer = clutter_seat_create_virtual_device (seat, CLUTTER_POINTER_DEVICE);
   now_us = g_get_monotonic_time ();
 
@@ -193,6 +199,12 @@ gesture_state_machine_move_to_cancelled_while_possible (void)
   gboolean was_updated;
   ClutterGestureState gesture_state_change;
 
+  if (g_getenv ("DEB_ALLOW_FLAKY_TESTS") == NULL)
+    {
+      g_test_skip ("https://gitlab.gnome.org/GNOME/mutter/-/issues/3521");
+      return;
+    }
+
   virtual_pointer = clutter_seat_create_virtual_device (seat, CLUTTER_POINTER_DEVICE);
   now_us = g_get_monotonic_time ();
 
@@ -240,6 +252,12 @@ gesture_state_machine_move_to_cancelled_on_sequence_cancel (void)
   gboolean was_updated;
   ClutterGestureState gesture_state_change;
 
+  if (g_getenv ("DEB_ALLOW_FLAKY_TESTS") == NULL)
+    {
+      g_test_skip ("https://gitlab.gnome.org/GNOME/mutter/-/issues/3521");
+      return;
+    }
+
   virtual_pointer = clutter_seat_create_virtual_device (seat, CLUTTER_POINTER_DEVICE);
   now_us = g_get_monotonic_time ();
 
@@ -297,6 +315,12 @@ gesture_multiple_mouse_buttons (void)
   ClutterGesture *gesture = CLUTTER_GESTURE (g_object_new (TEST_TYPE_GESTURE, NULL));
   gboolean was_updated;
 
+  if (g_getenv ("DEB_ALLOW_FLAKY_TESTS") == NULL)
+    {
+      g_test_skip ("https://gitlab.gnome.org/GNOME/mutter/-/issues/3521");
+      return;
+    }
+
   virtual_pointer = clutter_seat_create_virtual_device (seat, CLUTTER_POINTER_DEVICE);
   now_us = g_get_monotonic_time ();
 
diff --git a/src/tests/clutter/conform/grab.c b/src/tests/clutter/conform/grab.c
index 93a8cf4..9018a53 100644
--- a/src/tests/clutter/conform/grab.c
+++ b/src/tests/clutter/conform/grab.c
@@ -620,6 +620,12 @@ grab_input_only (void)
   ClutterSeat *seat;
   g_autoptr (ClutterVirtualInputDevice) pointer = NULL;
 
+  if (g_getenv ("DEB_ALLOW_FLAKY_TESTS") == NULL)
+    {
+      g_test_skip ("https://gitlab.gnome.org/GNOME/mutter/-/issues/3205");
+      return;
+    }
+
   seat = clutter_test_get_default_seat ();
   pointer = clutter_seat_create_virtual_device (seat, CLUTTER_POINTER_DEVICE);
 
diff --git a/src/tests/meson.build b/src/tests/meson.build
index 0b1c0f0..b72c209 100644
--- a/src/tests/meson.build
+++ b/src/tests/meson.build
@@ -423,6 +423,7 @@ test_cases += [
     'name': 'input-capture',
     'suite': 'backends/native',
     'sources': [ 'input-capture-tests.c' ],
+    'flaky': true,
   },
   {
     'name': 'thread',
@@ -714,8 +715,14 @@ foreach test_case: test_cases
 
   test_depends = [ default_plugin ] + test_case.get('depends', [])
 
+  extra_suites = []
+
+  if test_case.get('flaky', false)
+    extra_suites += 'flaky'
+  endif
+
   test(test_case['name'], test_executable,
-    suite: ['core', 'mutter/' + test_case['suite']],
+    suite: ['core', 'mutter/' + test_case['suite']] + extra_suites,
     env: test_env,
     depends: test_depends,
     is_parallel: false,
@@ -777,7 +784,7 @@ stacking_tests = [
 
 foreach stacking_test: stacking_tests
   test(stacking_test, test_runner,
-    suite: ['core', 'mutter/stacking'],
+    suite: ['core', 'mutter/stacking', 'flaky'],
     env: test_env,
     args: [
       files('stacking' / stacking_test + '.metatest'),
